trin-core/src/portalnet/events.rs

```diff
@@ -114,20 +112,93 @@ impl PortalnetEvents {
     // This could be handled in the UtpListener impl, but for consistency with data handling
     // and clarity it can be put here
     async fn process_utp_byte_stream(&mut self) {
-        for (_, conn) in self.utp_listener.utp_connections.iter_mut() {
+        for (_, conn) in self.utp_listener.write().await.utp_connections.iter_mut() {
             let received_stream = conn.recv_data_stream.clone();
-            if received_stream.is_empty() {
-                continue;
+
+            if UtpMessage::is_valid(&received_stream[..]) {
+                let message = UtpMessage::decode(&received_stream[..]);
+
+                match self
+                    .utp_listener
+                    .read()
+                    .await
+                    .listening
+                    .get(&conn.conn_id_recv)
+                {
+                    Some(message_type) => match message_type {
+                        // 8 is the Offer/Accept uTP stream handler
+                        8 => {
+                            let payload = UtpAccept::from_ssz_bytes(&message.payload[..])
+                                .unwrap()
+                                .message;
+
+                            for (key, content) in payload {
+                                store(key, content).unwrap();
+                            }
+                        }
+                        _ => warn!("Received uTP Stream Message of unknown type"),
+                    },
+                    _ => warn!("uTP listening HashMap doesn't have uTP stream message type"),
+                }
+
+                // message was successfully handled, remove it from the buffer
+                conn.recv_data_stream = received_stream[(message.len() as usize)..].to_owned();
             }
+        }
+    }
+}
 
-            let message_len = u32::from_be_bytes(received_stream[0..4].try_into().unwrap());
+pub struct UtpMessage {
```

Comment:
We can do that, I wonder what the file should be called since it shouldn't go into utp.rs and it could be confused for something uTP implementation wise. I will give it some thought

Reply:
I think it is probably worth having a separate folder or why not a crate for files related to the utp implementation.

Reply:
@ogenev UtpMessage has nothing to do with the uTP implementation. I mean we can move utp.rs to a crate. UtpMessage is just for handling uTP messages on Trin's side.

Reply:
Anyways I will make a PR for that I guess because Sam also made that suggestion

Reply:
Giving it some thought I will just make a folder. The uTP implementation is heavily tied to Discv5. Also this UtpMessage stuff isn't uTP implementation it is just a impl for cleaning up the handling of streams for Trin so it would probably make the most sense to just have it in a folder. I still need to think about what to call the new file uTP-handling I honestly have no idea.